Add windows support

laurentdesmet opened the pull request:
See #226

gigabo commented:
Looks like `cross-spawn-async` has been [superseded by `cross-spawn`](https://github.com/IndigoUnited/node-cross-spawn-async#cross-spawn-async):

> This module is deprecated, use [cross-spawn](https://github.com/IndigoUnited/node-cross-spawn) instead which no longer requires a build toolchain.

laurentdesmet commented:
Wow, I was looking for the difference between these two modules and I completely missed that... So I decided to take the one with the most downloads. Thanks for notifying!

Daniel15 commented:
Thanks! Did you test this on Windows?

laurentdesmet commented:
I did test it quickly by hand on a Windows machine and it seemed to work, but there is still some work to do for all the tests to pass on Windows. I'll fix the tests for Windows in one of the following days. Adding a Windows CI build using appveyor as in #230 is the next step to verify this automatically.

laurentdesmet commented:
I already made some progress making the tests pass on Windows using appveyor but I don't have access to a Windows machine right now, which makes testing difficult because of slow feedback cycles. I'll continue when I have access to a Windows machine where I can run the tests locally.

hzoo commented:
There's already a few prs for windows - #230 like you mentioned as well as https://github.com/lerna/lerna/pull/171, https://github.com/lerna/lerna/pull/141. I would just rebase those changes into yours

laurentdesmet commented:
Ok, I will do this and continue testing

zkochan commented:
Is it going to be merged? Lerna bootstrap is currently not usable on Windows

laurentdesmet commented:
I could not find any free time to work on this but I'll continue working on this PR, the moment my schedule allows it

zkochan commented:
I might be able to work on this today. Have you pushed your latest changes?

laurentdesmet commented:
I pushed everything, definitely check #171 as @hzoo already mentioned here. It already contains a lot of work but I didn't check yet how much it is still up to date with the HEAD of the master branch.

zkochan commented:
@laurentdesmet, I rebased your changes on master and #171 then fixed the tests that were not passing on my Windows 10. I am going to make a pull request to your fork now

laurentdesmet commented:
I merged the changes into the master branch of my fork, locally on my machine (Mac) all the tests passed for all Node versions (also v0.10.46), but they did fail on Travis and AppVeyor. I should configure my Windows machine (which is not used for months now...) to test it locally on Windows. But because the tests where passing on your machine, it seems something AppVeyor related makes the test fail.

zkochan commented:
AppVeyor is slower than Travis, so the 3 timeout issues should be fixed by simply increasing the timeouts for those tests

The 4th failed test seems to be Windows related because the line ending didn't match. It is weird though that on my machine that test passed.

laurentdesmet commented:
Ok, Travis tests do pass now (https://travis-ci.org/laurentdesmet/lerna/builds/144003698) but I didn't change anything, just made a separate branch. This nondeterministic test results are a little bit annoying... Gonna check if I can find the cause of them.

laurentdesmet commented:
I found out why the 4th test failed on your machine. On line 265 of `test/PublishCommand.js`, `\n` got replaced by `EOL` but this should remain `\n` because this should have the same contents as the `lerna.json`. On Windows you have probably `core.autocrlf` set to true, which causes the file to have its line endings being converted to CRLF and that made test pass on your system.

zkochan commented:
Good job! I am super happy the latest lerna will work on Windows with no more issues!

laurentdesmet commented:
If this PR gets merged, then #171 and #230 can also be closed because its work is merged in this PR.

thejameskyle commented on test/PublishCommand.js:
```diff
@@ -234,19 +235,19 @@ describe("PublishCommand", () => {
           { args: ["git checkout -- packages/*/package.json"] },
 
-          { args: ["npm dist-tag ls package-1"], returns: "lerna-temp: 1.0.0-alpha.81e3b443\nstable: 1.0.0" },
+          { args: ["npm dist-tag ls package-1"], returns: "lerna-temp: 1.0.0-canary.81e3b443" + EOL + "stable: 1.0.0" },
```
Why is alpha being turned into canary. We definitely do not want canary #166

hzoo replied:
does this need to be rebased?

jdanyow commented:
will this be merged?

hzoo commented:
Yeah there was https://github.com/lerna/lerna/pull/236#discussion-diff-70688445 so need to look at the test output more carefully- maybe it just needs to be rebased?

laurentdesmet commented:
I rebased my master branch on top of the lerna:master branch

laurentdesmet commented:
aaa4cdb3e578353b75c4ce80349ce4bf0983fdb2 replaced alpha by canary, but I see no reason why, so I reverted this change. I also added the node versions 0.10, 1.12, 5 and 6 to the appveyor CI configuration file. 

I think there is one more thing we can do to polish the PR a little bit more: now we put `os.EOL` everywhere instead of `\n` and I think that we could maybe replace it by something like [normalize-newline](https://www.npmjs.com/package/normalize-newline) so we can still use `\n` instead of concatenating strings with the less readable `os.EOL`. I'll look into this during the weekend.

gigabo commented:
Hi @laurentdesmet - I just pushed a branch with this PR rebased onto current master, and [appveyor failed on node 0.12](https://ci.appveyor.com/project/hzoo/lerna/build/1.0.19).  Looks like it didn't make it past `npm install`.  Any idea what might be going on there?

laurentdesmet commented:
@gigabo I rebased this branch on top of master (that was updated with another two commits since your rebase) and it succeeds now. So no idea what the problem was, didn't check the changes from the last two commits on master.

gigabo commented:
Great!  Must have been a spurious failure... thanks for getting up to date.

The `normalize-newline` thing might be nice, but seems worth getting this merged in first.

So, this looks good to me.  @hzoo, @thejameskyle?

hzoo commented:
Yeah lets do it in a follow up pr then? It's been a too long for these changes so we should get them in
